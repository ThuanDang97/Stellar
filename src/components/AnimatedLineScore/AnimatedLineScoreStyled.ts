import styled, { keyframes } from 'styled-components'

// Components
import {
  EndBaseLineStyled,
  MiddleBaseLineStyled,
  StartBaseLineStyled,
} from '@components/styled-components/LineStyled'

// Utils
import { pxToRem } from '@utils/theme'

const fadeIn = keyframes`
  0% {
    opacity: 0.3;
  }
  30%, 100% {
    opacity: 1;
  }
`

const showLayer = (width: string, left: number) => keyframes`
  0% {
    width: 0;
    left: -${pxToRem(left)};
  }
  100% {
    width: calc(${width});
    left: 0
  }
`

export const StartLineScoreStyled = styled(StartBaseLineStyled)`
  animation: ${fadeIn} 0.6s linear ${({ delay }) => `${delay}s`} both;
`
export const MiddleLineScoreStyled = styled(MiddleBaseLineStyled)`
  margin-left: -${({ theme }) => pxToRem(theme.metrics.borderWidth.xl)};

  animation: ${fadeIn} 0.6s linear ${({ delay }) => `${delay}s`} both;

  &:nth-child(4) {
    position: relative;
    z-index: 10;
  }
`
export const EndLineScoreStyled = styled(EndBaseLineStyled)`
  position: relative;
  margin-left: -${({ theme }) => pxToRem(theme.metrics.borderWidth.xl)};

  animation: ${fadeIn} 0.6s linear ${({ delay }) => `${delay}s`} both;
  z-index: 10;
`

export const LayerLineStyled = styled(StartBaseLineStyled)`
  position: absolute;
  width: ${({ width }) => `calc(${width})`};
  height: ${({ theme }) => pxToRem(theme.metrics.borderWidth.xl)};

  border-top-color: ${({ theme }) => theme.colors.caribbeanGreen};
  border-right-color: ${({ theme }) => theme.colors.caribbeanGreen};
  background: ${({ theme }) => theme.colors.caribbeanGreen};

  animation: ${({ width, theme }) =>
      typeof width === 'string' &&
      showLayer(width, theme.metrics.borderWidth.xl)}
    1s ease-out both ${({ delay }) => `${(delay ?? 0) + 7.8}s`};
  z-index: 1;
`

export const Shim = styled.div<{ left: string }>`
  position: absolute;
  left: ${({ left }) => `calc(${left})`};

  width: ${({ theme }) => pxToRem(theme.metrics.dimensions.xs)};
  height: ${({ theme }) => pxToRem(theme.metrics.dimensions.sm)};
  transform: translateX(-7px) skew(-45deg);

  background-color: ${({ theme }) => theme.colors.bigStone};
  z-index: 100;
`
