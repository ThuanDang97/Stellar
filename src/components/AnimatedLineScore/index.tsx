import React, { memo } from 'react'
import { useTheme } from 'styled-components'

// Components
import { ViewStyled } from '@components/styled-components'
import {
  StartLineScoreStyled,
  MiddleLineScoreStyled,
  EndLineScoreStyled,
  LayerLineStyled,
  Shim,
} from '@components/AnimatedLineScore/AnimatedLineScoreStyled'

// Utils
import { pxToRem } from '@utils/theme'

// Props type
import { AnimationComponentProps } from '@self-types/components/Animation.props'

const AnimatedLineScore = ({ delay = 0 }: AnimationComponentProps) => {
  const theme = useTheme()

  return (
    <ViewStyled
      position="relative"
      display="flex"
      width={theme.metrics.width.full}
      alignItems="flex-start"
      overflow="hidden"
      mLeft={theme.metrics.dimensions.base}
    >
      {/* Five white lines */}
      <StartLineScoreStyled width={35} delay={0.5 + delay} />
      <MiddleLineScoreStyled width={30} delay={1.3 + delay} />
      <MiddleLineScoreStyled width={15} delay={2.1 + delay} />
      <MiddleLineScoreStyled width={10} delay={2.9 + delay} />
      <EndLineScoreStyled width={10} delay={3.7 + delay} />

      {/* CaribbeanGreen line */}
      <LayerLineStyled width={`80% - ${pxToRem(14)}`} delay={delay} />

      {/* Shim between lines */}
      <Shim left="35%" />
      <Shim left={`65% - ${pxToRem(7)}`} />
      <Shim left={`80% - ${pxToRem(14)}`} />
      <Shim left={`90% - ${pxToRem(21)}`} />
    </ViewStyled>
  )
}

export default memo(AnimatedLineScore)
