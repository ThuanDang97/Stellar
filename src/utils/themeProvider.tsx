import { ThemeProvider } from 'styled-components'
import { render } from '@testing-library/react'
import { theme } from '@themes/index'

export const renderWithTheme = (component: React.ReactNode) => {
  return render(<ThemeProvider theme={theme}>{component}</ThemeProvider>)
}
