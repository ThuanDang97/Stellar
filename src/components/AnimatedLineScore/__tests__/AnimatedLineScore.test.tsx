// Components
import { renderWithTheme } from '@utils/themeProvider'
import AnimatedLineScore from '..'

describe('AnimatedLineScore Component', () => {
  it('renders animatedLineScore unchanged', () => {
    const { asFragment } = renderWithTheme(<AnimatedLineScore />)

    expect(asFragment()).toMatchSnapshot()
  })
})
