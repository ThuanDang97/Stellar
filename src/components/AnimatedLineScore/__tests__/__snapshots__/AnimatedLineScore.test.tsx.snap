// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AnimatedLineScore Component renders animatedLineScore unchanged 1`] = `
<DocumentFragment>
  <div
    class="ViewStyled-sc-rqe5sa-0 kCJugR"
    display="flex"
    overflow="hidden"
    width="100%"
  >
    <div
      class="LineStyled__BaseLineStyled-sc-okjczv-0 LineStyled__StartBaseLineStyled-sc-okjczv-1 AnimatedLineScoreStyled__StartLineScoreStyled-sc-poybgv-0 dlHdcM hLygIM dPidGS"
      width="35"
    />
    <div
      class="LineStyled__BaseLineStyled-sc-okjczv-0 LineStyled__MiddleBaseLineStyled-sc-okjczv-2 AnimatedLineScoreStyled__MiddleLineScoreStyled-sc-poybgv-1 btYAst ldvwuf fvdruU"
      width="30"
    />
    <div
      class="LineStyled__BaseLineStyled-sc-okjczv-0 LineStyled__MiddleBaseLineStyled-sc-okjczv-2 AnimatedLineScoreStyled__MiddleLineScoreStyled-sc-poybgv-1 dCqiri ldvwuf hkQYyp"
      width="15"
    />
    <div
      class="LineStyled__BaseLineStyled-sc-okjczv-0 LineStyled__MiddleBaseLineStyled-sc-okjczv-2 AnimatedLineScoreStyled__MiddleLineScoreStyled-sc-poybgv-1 hhHjIH ldvwuf ifzRSN"
      width="10"
    />
    <div
      class="LineStyled__BaseLineStyled-sc-okjczv-0 LineStyled__EndBaseLineStyled-sc-okjczv-3 AnimatedLineScoreStyled__EndLineScoreStyled-sc-poybgv-2 hhHjIH dahDgB jlmGRx"
      width="10"
    />
    <div
      class="LineStyled__BaseLineStyled-sc-okjczv-0 LineStyled__StartBaseLineStyled-sc-okjczv-1 AnimatedLineScoreStyled__LayerLineStyled-sc-poybgv-3 kdqgjs hLygIM gwDPoj"
      width="80% - 0.875rem"
    />
    <div
      class="AnimatedLineScoreStyled__Shim-sc-poybgv-4 lisOCr"
    />
    <div
      class="AnimatedLineScoreStyled__Shim-sc-poybgv-4 bDcpsq"
    />
    <div
      class="AnimatedLineScoreStyled__Shim-sc-poybgv-4 cRdoku"
    />
    <div
      class="AnimatedLineScoreStyled__Shim-sc-poybgv-4 dnSNqt"
    />
  </div>
</DocumentFragment>
`;
